File: src/models/board.rs
use crate::models::location::BOARD_CELLS;
use crate::*;

#[derive(Clone)]
pub(crate) struct Board {
    pub(crate) items: [Option<Item>; BOARD_CELLS],
    variant: GameVariant,
}

impl std::fmt::Debug for Board {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let occupied: Vec<(Location, &Item)> = self
            .items
            .iter()
            .enumerate()
            .filter_map(|(idx, opt)| opt.as_ref().map(|item| (Location::from_index(idx), item)))
            .collect();
        f.debug_struct("Board")
            .field("variant", &self.variant)
            .field("items", &occupied)
            .finish()
    }
}

impl PartialEq for Board {
    fn eq(&self, other: &Self) -> bool {
        self.items == other.items && self.variant == other.variant
    }
}

impl Eq for Board {}

impl Default for Board {
    fn default() -> Self {
        Self::new()
    }
}

impl Board {
    pub fn new() -> Self {
        Self::new_with_variant(GameVariant::DEFAULT)
    }

    pub fn new_with_variant(variant: GameVariant) -> Self {
        Self::from_items_array(Config::initial_items_array_for_variant(variant), variant)
    }

    pub(crate) fn from_items_array(
        items: [Option<Item>; BOARD_CELLS],
        variant: GameVariant,
    ) -> Self {
        Self { items, variant }
    }

    #[inline]
    pub fn remove_item(&mut self, location: Location) {
        self.items[location.index()] = None;
    }

    #[inline]
    pub fn put(&mut self, item: Item, location: Location) {
        self.items[location.index()] = Some(item);
    }

    #[inline]
    pub fn item(&self, location: Location) -> Option<&Item> {
        if !location.is_valid() {
            return None;
        }
        self.items[location.index()].as_ref()
    }

    #[inline]
    pub fn square(&self, location: Location) -> Square {
        if !location.is_valid() {
            return Square::Regular;
        }
        Config::square_at_for_variant(location, self.variant)
    }

    #[inline]
    pub fn variant(&self) -> GameVariant {
        self.variant
    }

    #[cfg(any(target_arch = "wasm32", test))]
    pub fn all_mons_bases(&self) -> Vec<Location> {
        Config::MONS_BASE_LOCATIONS.to_vec()
    }

    #[inline]
    pub fn supermana_base(&self) -> Location {
        Config::SUPERMANA_BASE
    }

    pub fn all_mons_locations(&self, color: Color) -> Vec<Location> {
        self.items
            .iter()
            .enumerate()
            .filter_map(|(idx, opt)| {
                if let Some(item) = opt {
                    if let Some(mon) = item.mon() {
                        if mon.color == color {
                            return Some(Location::from_index(idx));
                        }
                    }
                }
                None
            })
            .collect()
    }

    pub fn all_free_regular_mana_locations(&self, color: Color) -> Vec<Location> {
        self.items
            .iter()
            .enumerate()
            .filter_map(|(idx, opt)| match opt {
                Some(Item::Mana {
                    mana: Mana::Regular(mana_color),
                }) if *mana_color == color => Some(Location::from_index(idx)),
                _ => None,
            })
            .collect()
    }

    pub fn base(&self, mon: Mon) -> Location {
        Config::mon_base(mon.kind, mon.color)
    }

    pub fn fainted_mons_locations(&self, color: Color) -> Vec<Location> {
        self.items
            .iter()
            .enumerate()
            .filter_map(|(idx, opt)| match opt {
                Some(Item::Mon { mon }) if mon.color == color && mon.is_fainted() => {
                    Some(Location::from_index(idx))
                }
                _ => None,
            })
            .collect()
    }

    pub fn find_mana(&self, color: Color) -> Option<Location> {
        self.items
            .iter()
            .enumerate()
            .find_map(|(idx, opt)| match opt {
                Some(Item::Mana {
                    mana: Mana::Regular(mana_color),
                }) if *mana_color == color => Some(Location::from_index(idx)),
                _ => None,
            })
    }

    pub fn find_awake_angel(&self, color: Color) -> Option<Location> {
        self.items.iter().enumerate().find_map(|(idx, opt)| {
            if let Some(item) = opt {
                if let Some(mon) = item.mon() {
                    if mon.color == color && mon.kind == MonKind::Angel && !mon.is_fainted() {
                        return Some(Location::from_index(idx));
                    }
                }
            }
            None
        })
    }

    /// Iterate over occupied cells: yields (Location, &Item)
    #[inline]
    #[cfg(any(target_arch = "wasm32", test))]
    pub fn occupied(&self) -> impl Iterator<Item = (Location, &Item)> {
        self.items
            .iter()
            .enumerate()
            .filter_map(|(idx, opt)| opt.as_ref().map(|item| (Location::from_index(idx), item)))
    }
}

File: src/engine/board.ts
import {
  Color,
  MonKind,
  cloneItem,
  isMonFainted,
  itemEquals,
  itemMon,
  type Item,
  type Mon,
  type Square,
} from "./domain.js";
import {
  DEFAULT_GAME_VARIANT,
  GameVariant,
  MON_BASE_LOCATIONS,
  SUPERMANA_BASE,
  initialItemsForVariant,
  monBase,
  squareAtForVariant,
} from "./config.js";
import {
  ALL_LOCATIONS,
  BOARD_CELLS,
  checkedWrappedLocationIndex,
  cloneLocation,
  fromLocationIndex,
  isValidLocation,
  locationIndex,
  type Location,
} from "./geometry.js";

const BOARD_ITEMS_INITIALIZATION = Symbol();

type OccupiedBoardEntry = readonly [Location, Item];

type BoardItemsInitialization = {
  readonly [BOARD_ITEMS_INITIALIZATION]:
    | {
        readonly kind: "deep";
        readonly items: readonly (Item | undefined)[];
      }
    | {
        readonly kind: "simulation";
        readonly items: readonly (Item | undefined)[];
        readonly occupiedEntries: readonly OccupiedBoardEntry[];
      };
};

function cloneBoardItems(
  items: readonly (Item | undefined)[],
): (Item | undefined)[] {
  return Array.from({ length: BOARD_CELLS }, (_, index) => {
    const item = items[index];
    return item === undefined ? undefined : cloneItem(item);
  });
}

export class Board {
  private readonly itemSlots: (Item | undefined)[];
  private occupiedEntries: readonly OccupiedBoardEntry[] | undefined;
  readonly #gameVariant: GameVariant;

  public constructor(variant?: GameVariant);
  /** @internal Direct initialization for value-preserving board copies. */
  public constructor(
    variant: GameVariant,
    // eslint-disable-next-line @typescript-eslint/unified-signatures -- This internal overload is stripped from declarations, preserving the public constructor signature.
    initialization: BoardItemsInitialization,
  );
  public constructor(
    variant: GameVariant = DEFAULT_GAME_VARIANT,
    initialization?: BoardItemsInitialization,
  ) {
    const value = initialization?.[BOARD_ITEMS_INITIALIZATION];
    this.itemSlots =
      value === undefined
        ? initialItemsForVariant(variant)
        : value.kind === "deep"
          ? cloneBoardItems(value.items)
          : [...value.items];
    this.#gameVariant = variant;
    this.occupiedEntries =
      value?.kind === "simulation" ? value.occupiedEntries : undefined;
  }

  public get items(): readonly (Item | undefined)[] {
    return [...this.itemSlots];
  }

  public static fromItems(
    items: readonly (Item | undefined)[],
    variant: GameVariant,
  ): Board {
    return new Board(variant, {
      [BOARD_ITEMS_INITIALIZATION]: { kind: "deep", items },
    });
  }

  public clone(): Board {
    return Board.fromItems(this.itemSlots, this.#gameVariant);
  }

  /** @internal Shallow value copy for trusted simulation paths. */
  public cloneForSimulation(): Board {
    return new Board(this.#gameVariant, {
      [BOARD_ITEMS_INITIALIZATION]: {
        kind: "simulation",
        items: this.itemSlots,
        occupiedEntries: this.getOccupiedEntries(),
      },
    });
  }

  public removeItem(at: Location): void {
    this.itemSlots[checkedWrappedLocationIndex(at)] = undefined;
    this.occupiedEntries = undefined;
  }

  /** Store a value-copy, mirroring Rust's Copy item argument. */
  public put(item: Item, at: Location): void {
    this.itemSlots[checkedWrappedLocationIndex(at)] = cloneItem(item);
    this.occupiedEntries = undefined;
  }

  /** Internal live item reference. Public wrappers must clone before returning it. */
  public item(at: Location): Item | undefined {
    if (!isValidLocation(at)) {
      return undefined;
    }
    return this.itemSlots[locationIndex(at)];
  }

  public square(at: Location): Square {
    return isValidLocation(at)
      ? squareAtForVariant(at, this.#gameVariant)
      : { kind: "regular" };
  }

  public variant(): GameVariant {
    return this.#gameVariant;
  }

  public allMonsBases(): Location[] {
    return MON_BASE_LOCATIONS.map(cloneLocation);
  }

  public supermanaBase(): Location {
    return cloneLocation(SUPERMANA_BASE);
  }

  public allMonsLocations(color: Color): Location[] {
    const result: Location[] = [];
    for (let index = 0; index < BOARD_CELLS; index += 1) {
      const item = this.itemSlots[index];
      if (item !== undefined && itemMon(item)?.color === color) {
        result.push(fromLocationIndex(index));
      }
    }
    return result;
  }

  public allFreeRegularManaLocations(color: Color): Location[] {
    const result: Location[] = [];
    for (let index = 0; index < BOARD_CELLS; index += 1) {
      const item = this.itemSlots[index];
      if (
        item?.kind === "mana" &&
        item.mana.kind === "regular" &&
        item.mana.color === color
      ) {
        result.push(fromLocationIndex(index));
      }
    }
    return result;
  }

  public base(mon: Mon): Location {
    return cloneLocation(monBase(mon.kind, mon.color));
  }

  public faintedMonsLocations(color: Color): Location[] {
    const result: Location[] = [];
    for (let index = 0; index < BOARD_CELLS; index += 1) {
      const item = this.itemSlots[index];
      if (
        item?.kind === "mon" &&
        item.mon.color === color &&
        isMonFainted(item.mon)
      ) {
        result.push(fromLocationIndex(index));
      }
    }
    return result;
  }

  public findMana(color: Color): Location | undefined {
    for (let index = 0; index < BOARD_CELLS; index += 1) {
      const item = this.itemSlots[index];
      if (
        item?.kind === "mana" &&
        item.mana.kind === "regular" &&
        item.mana.color === color
      ) {
        return fromLocationIndex(index);
      }
    }
    return undefined;
  }

  public findAwakeAngel(color: Color): Location | undefined {
    for (let index = 0; index < BOARD_CELLS; index += 1) {
      const item = this.itemSlots[index];
      const mon = item === undefined ? undefined : itemMon(item);
      if (
        mon?.color === color &&
        mon.kind === MonKind.Angel &&
        !isMonFainted(mon)
      ) {
        return fromLocationIndex(index);
      }
    }
    return undefined;
  }

  /** @internal Compare board storage without allocating public snapshots. */
  public itemsEqual(other: Board): boolean {
    for (let index = 0; index < BOARD_CELLS; index += 1) {
      const leftItem = this.itemSlots[index];
      const rightItem = other.itemSlots[index];
      if (leftItem === undefined || rightItem === undefined) {
        if (leftItem !== rightItem) {
          return false;
        }
      } else if (!itemEquals(leftItem, rightItem)) {
        return false;
      }
    }
    return true;
  }

  private getOccupiedEntries(): readonly OccupiedBoardEntry[] {
    if (this.occupiedEntries !== undefined) {
      return this.occupiedEntries;
    }
    const occupiedEntries: OccupiedBoardEntry[] = [];
    for (let index = 0; index < BOARD_CELLS; index += 1) {
      const item = this.itemSlots[index];
      if (item !== undefined) {
        const at = ALL_LOCATIONS[index];
        if (at !== undefined) {
          occupiedEntries.push([at, item]);
        }
      }
    }
    this.occupiedEntries = occupiedEntries;
    return occupiedEntries;
  }

  /** Iterate a snapshot of the board's occupied locations in row-major order. */
  public occupied(): IterableIterator<readonly [Location, Item]> {
    return this.getOccupiedEntries().values();
  }
}

export function boardEquals(left: Board, right: Board): boolean {
  if (left.variant() !== right.variant()) {
    return false;
  }
  return left.itemsEqual(right);
}
